### PR TITLE
[vcpkg-scripts][boost] Improve generate-ports.ps1

### PR DESCRIPTION
**Pull request by dg0yt**
- Change homepage URLs to point to documentation instead of GH.
 - Allow encoding alternative propagation of `supports`/`platform` so that changes don't need to be selected manually.
   - Allow to suppress the generated dependency `platform` expression when a dependency is non-optional.
     (`boost-parameter` requires `boost-python`, and the dependency transitively determines the supported platforms of the dependent port.)
   - For port `boost`, generate `platform` expression from transitive `supports` limitations.  
     (`boost-parameter` platform expression must account for `boost-python`.)
     This fixes port `boost` for uwp. And so it can now be directly referenced from `vcpkg-ci-boost` which is already enforced to  `pass` in ci.baseline.txt. :tada: 
 - Restores sorting of `$portData`.
 - Use the same tarball download area and naming as `vcpkg install`.
 
This PR doesn't include the updates to `boost-container` (#38806), `bost-interprocess` (#38815) and boost-math (#38728)

### CODE REVIEW

**Comment by dg0yt** [comment]
FTR I do have a variant of `boost-generate.ps1` that parses `CMakelists.txt` for use `Boost::` targets instead of headers and sources for use of  `#include`. With all tarballs downloaded, generation run time is down from 180 s to 6 s. However, the list of dependencies is significantly different.
 - Addition of dependencies due to use of unneeded CMake targets (cf. #38806)
 - Addition of dependencies due to being unable to detect conditions. (However, the current parser isn't aware of conditions either.)
 - Removal of dependencies which I assume to be still available as transitive dependencies.

File: scripts/boost/generate-ports.ps1 [hunk 1]
```diff
@@ -284,7 +316,7 @@ function GeneratePort() {
     # Generate vcpkg.json
     GeneratePortManifest `
         -PortName $portName `
-        -Homepage "https://github.com/boostorg/$Library" `
+        -Homepage "https://www.boost.org/libs/$Library" `
```
Comment (dg0yt) [inline scripts/boost/generate-ports.ps1:319]: This is what the GH repos use, e.g. https://github.com/boostorg/asio.

**Comment by Kojoley** [comment]
> FTR I do have a variant of `boost-generate.ps1` that parses `CMakelists.txt` for use of `Boost::` targets instead of parsing headers and sources for use of `#include`. With all tarballs downloaded, generation run time is down from 180 s to 6 s. However, the list of dependencies is significantly different.

That should've been done when build was changed to cmake. A better solution would be to ask Boost to list dependencies in `meta/libraries.json`.

**Comment by dg0yt** [comment]
> That should've been done when build was changed to cmake. A better solution would be to ask Boost to list dependencies in `meta/libraries.json`.

Again, the key problem is to attach the right guards on conditional dependencies.

**Comment by jimwang118** [comment]
Add reviewed tag for further review by the vcpkg team.

**Review by vicroms — DISMISSED** [review]
LGTM, I'll ask for a second review in tomorrow's meeting.

**Comment by Neumann-A** [comment]
WAH why is the script flashing terminals in my face now.

File: scripts/boost/generate-ports.ps1 [hunk 2]
```diff
@@ -167,14 +176,36 @@ function GeneratePortName() {
...
+    elseif ($portData.Contains($PortName) -and $portData[$PortName].Contains('supports')) {
+        @{name = $PortName; platform = $portData[$PortName]['supports'] }
+    }
+    elseif ($ForLibrary -eq '' -and $suppressPlatformForDependency.Contains($PortName)) {
+        # For 'boost'.
+        $platform = `
+            $suppressPlatformForDependency[$PortName] `
+            | ForEach-Object { (GeneratePortDependency -PortName $_).platform } `
+            | Group-Object -NoElement `
+            | Join-String -Property Name -Separator ' & '
```
Comment (dg0yt) [inline scripts/boost/generate-ports.ps1:199]: May need to wrap terms in `(...)` before joining with ` & `.

**Pull request merged**